### ISSUE TITLE
SoundCloud 'Discover' feature freezes my entire system

### ISSUE THREAD

**Issue by CrazyLemon** [tiliado/nuvolaplayer#245]
**A workaround for this issue has landed n the [master](https://github.com/tiliado/nuvolaplayer) & [3.0.x](https://github.com/tiliado/nuvolaplayer/tree/3.0.x) git branches. Thanks for your bug report.**

At first, this fix will be available with the next rebuilt of packages of the [Nuvola Player 3.1 Rolling Releases](https://tiliado.github.io/nuvolaplayer/documentation/3.1/install.html). Later, the fix will be also included in the next bug-fix release of the [Nuvola Player 3.0 Stable Releases](https://tiliado.github.io/nuvolaplayer/documentation/3.0/install.html). To get informed when Nuvola Player 3.0.4 Stable Release is released, follow Nuvola Player on [Facebook](https://www.facebook.com/nuvolaplayer), [Google+](https://plus.google.com/110794636546911932554) or [Twitter](https://twitter.com/NuvolaPlayer) or subscribe to the Nuvola Player Newsletter: [weekly](http://eepurl.com/bLbm5H) or [monthly](http://eepurl.com/bLbtM1).

**Update:** This issue has been fixed in the [Nuvola Player 3.0.4 Stable Releases](https://tiliado.github.io/nuvolaplayer/documentation/3.0/install.html) (September 17th, 2016).

---

**Original report:**

So since introduction of 'Discover' feature in soundcloud and well in nuvola player as part of it it has been freezing my system entirly.
By freeze i mean no keyboard response, no mouse response (sometimes, sometimes there is).
As soon as i clicked on 'Discover' everything froze.
This has been happening on Ubuntu 16.04. I can't give you the exact nuvolaplayer version because it has been updated since then.

Since the update i was first 'warned' that i should use newer webkit version. Hurray i said to myself. No more freezes. I added the repo and updated the webkit version.
But freezes kept occuring. This time i could partially use Discover feature, but it was slow and graphics were (sometimes) weird.

It froze right now as i tried to run it from the terminal in hope i could get you some debugging info but nothing useful came out. Current info

> nuvolaplayer3 --version
> Nuvola Player 3 3.0.3
> Revision 8e3fd1301c4f055128063cce690165e832429265
> WebKitGTK 2.12.2
> libsoup 2.52.2
This is all output i could get when nuvolaplayer is running 'properly'

>  nuvolaplayer3
> Runner: [WARNING  GLib-GObject] g_object_set_property: construct property "local-storage-directory" for object 'WebKitWebContext' can't be set after construction
> Worker: [DEBUG    Nuvola] WebExtension.vala:220: Page 1 created for (null)
> Worker: [DEBUG    Nuvola] WebExtension.vala:291: Document loaded about:loading
> Worker: [DEBUG    Nuvola] WebExtension.vala:130: Window object cleared for 'https://soundcloud.com/stream'
> Worker: [DEBUG    Nuvola] WebExtension.vala:140: Init frame: https://soundcloud.com/stream, 0x55a0bc08d210, 0x55a0bb92eba0, 0x7fee48363140
> Runner: [CRITICAL  Json] json_node_get_parent: assertion 'JSON_NODE_IS_VALID (node)' failed
> Runner: [CRITICAL  Json] Application will not function properly.
> Runner: [CRITICAL  GLib] g_ptr_array_remove_index: assertion 'index_ < rarray->len' failed
> Runner: [CRITICAL  GLib] Application will not function properly.
> Runner: [CRITICAL  Json] json_node_get_parent: assertion 'JSON_NODE_IS_VALID (node)' failed
> Runner: [CRITICAL  Json] Application will not function properly.
> Runner: [CRITICAL  GLib] g_ptr_array_remove_index: assertion 'index_ < rarray->len' failed
> Runner: [CRITICAL  GLib] Application will not function properly.
> Runner: https://a-v2.sndcdn.com/assets/app-90ef4-c7a52dc.js:34:7488: CONSOLE INFO You like to look under the hood? Why not help us build the engine? http://soundcloud.com/jobs
> Worker: [DEBUG    Nuvola] WebExtension.vala:291: Document loaded https://soundcloud.com/stream
> Runner: https://www.googleadservices.com/pagead/conversion.js:20:210: CONSOLE ERROR Blocked a frame with origin "https://5485101.fls.doubleclick.net" from accessing a frame with origin "https://soundcloud.com". Protocols, domains, and ports must match.
> Runner: https://www.googleadservices.com/pagead/conversion.js:20:258: CONSOLE ERROR Blocked a frame with origin "https://5485101.fls.doubleclick.net" from accessing a frame with origin "https://soundcloud.com". Protocols, domains, and ports must match.
> Runner: https://www.googleadservices.com/pagead/conversion.js:20:210: CONSOLE ERROR Blocked a frame with origin "https://5485101.fls.doubleclick.net" from accessing a frame with origin "https://soundcloud.com". Protocols, domains, and ports must match.
> https://www.googleadservices.com/pagead/conversion.js:20:258: CONSOLE ERROR Blocked a frame with origin "https://5485101.fls.doubleclick.net" from accessing a frame with origin "https://soundcloud.com". Protocols, domains, and ports must match.
> Runner: [WARNING  Nuvola] AppRunnerController.vala:705: Communication failed: Failed to read message size.
> Runner: [WARNING  Nuvola] AppRunnerController.vala:705: Communication failed: Failed to read from socket. Napaka med prejemanjem podatkov: Connection reset by peer
> Runner: [WARNING  Nuvola] AppRunnerController.vala:725: Communication failed: Failed to connect to socket '/tmp/libdiorite.nuvolaplayer3.soundcloud.webworker(crazylemon)'. Connection refused
> Runner: [WARNING  Nuvola] AppRunnerController.vala:705: Communication failed: Failed to connect to socket '/tmp/libdiorite.nuvolaplayer3.soundcloud.webworker(ccrazylemon)'. Connection refused
> Runner: [WARNING  Nuvola] AppRunnerController.vala:705: Communication failed: Failed to connect to socket '/tmp/libdiorite.nuvolaplayer3.soundcloud.webworker(crazylemon)'. Connection refused
> Runner: [WARNING  Nuvola] AppRunnerController.vala:725: Communication failed: Failed to connect to socket '/tmp/libdiorite.nuvolaplayer3.soundcloud.webworker(crazylemon)'. Connection refused
> Runner: [WARNING  Nuvola] AppRunnerController.vala:705: Communication failed: Failed to connect to socket '/tmp/libdiorite.nuvolaplayer3.soundcloud.webworker(crazylemon)'. Connection refused
> Runner: [WARNING  Nuvola] AppRunnerController.vala:705: Communication failed: Failed to connect to socket '/tmp/libdiorite.nuvolaplayer3.soundcloud.webworker(crazylemon)'. Connection refused
> Runner: [WARNING  Nuvola] AppRunnerController.vala:725: Communication failed: Failed to connect to socket '/tmp/libdiorite.nuvolaplayer3.soundcloud.webworker(crazylemon)'. Connection refused
> Worker: [DEBUG    Nuvola] WebExtension.vala:220: Page 1 created for (null)
> Worker: [CRITICAL Nuvola] nuvola_js_executor_call_function: assertion 'self != NULL' failed
> Worker: [CRITICAL Nuvola] Application will not function properly.
> Worker: [DEBUG    Nuvola] WebExtension.vala:130: Window object cleared for 'https://soundcloud.com/stream'
> Worker: [DEBUG    Nuvola] WebExtension.vala:140: Init frame: https://soundcloud.com/stream, 0x561decb0b500, 0x561decb31920, 0x7fbd29bdf140
> Worker: [CRITICAL  GLib] g_hash_table_insert_internal: assertion 'hash_table != NULL' failed
> Worker: [CRITICAL  GLib] Application will not function properly.
> Worker: [CRITICAL Nuvola] nuvola_js_api_inject: assertion 'self != NULL' failed
> Worker: [CRITICAL Nuvola] Application will not function properly.
> Worker: [CRITICAL Nuvola] nuvola_js_api_integrate: assertion 'self != NULL' failed
> Worker: [CRITICAL Nuvola] Application will not function properly.
> Worker: [CRITICAL Nuvola] nuvola_js_executor_call_function: assertion 'self != NULL' failed
> Worker: [CRITICAL Nuvola] Application will not function properly.
> Worker: [CRITICAL Nuvola] nuvola_js_executor_call_function: assertion 'self != NULL' failed
> Worker: [CRITICAL Nuvola] Application will not function properly.

**Comment by fenryxo** [comment]
Hello @CrazyLemon. Thanks for taking the time to report this issue. Which graphic card and driver do you use?

**Comment by CrazyLemon** [comment]
> *-display               
>        description: VGA compatible controller
>        product: BeaverCreek [Radeon HD 6530D]
>        vendor: Advanced Micro Devices, Inc. [AMD/ATI]
>        physical id: 1
>        bus info: pci@0000:00:01.0
>        version: 00
>        width: 32 bits
>        clock: 33MHz
>        capabilities: vga_controller bus_master cap_list rom
>        configuration: driver=radeon latency=0
> 
In short I'm using an APU so the integrated card is Radeon HD 6530D and the driver is the open source one - radeon.

One offtopic note... i keep getting white screen on the normal 'Stream' too. Debugging says something about unable to connect to a socket.. should open a new issue or is this the same sympthome?

**Comment by fenryxo** [comment]
> unable to connect to a socket

That means that the web worker process crashed and it is not possible to communicate with it.

> should i open a new issue or is this the same sympthome?

It's the same.

Is the freeze/crash of Soundcloud web page reproducible in [WebKitGTK MiniBrowser](https://github.com/tiliado/nuvolaplayer/wiki/Bug-Reporting-Guidelines#webkitgtk-minibrowser)?.

**Comment by CrazyLemon** [comment]
![img_20160807_163748](https://cloud.githubusercontent.com/assets/3321382/17463105/8e8fafce-5cbf-11e6-884d-5ed26b900952.jpg)

So a bit of explanation of the image. I took the image because the system froze..again.
First few bits and pieces are from Stream browsing around. 
First message is 
"radeon: Not enough memory for command submission"
the second and third are 
"Error sending IPC message: Conn" - presumably Connection reset by peer and
"Error sending IPC message: Broken pipe"
after that ^T i started using Discover and got weird graphic issues with that horizontal scroll hence the radeon message (same message as the first one).
And then my system froze. SSH was still responsive, mouse was responsive untill i killed minibrowser.
Webkitprocess was 'defunct'ed.

**Comment by fenryxo** [comment]
Well, the issue goes pretty deep. Since the issue is reproducible in the WebKitGTK MiniBrowser, I believe there is a bug in WebKitGTK (or deeper in the graphics stack) and not in Nuvola Player. Could you [report the issue upstream to the WebKit bug tracker](https://bugs.webkit.org/enter_bug.cgi?assigned_to=webkit-unassigned%40lists.webkit.org&attachurl=&blocked=&bug_file_loc=http://&bug_severity=Normal&bug_status=NEW&comment=&component=WebKit%20Gtk&contenttypeentry=&contenttypemethod=autodetect&contenttypeselection=text/plain&data=&dependson=&description=&flag_type-1=X&flag_type-3=X&flag_type-4=X&form_name=enter_bug&keywords=GTK&maketemplate=Remember%20values%20as%20bookmarkable%20template&op_sys=Linux&priority=P3&product=WebKit&rep_platform=PC&short_desc=%5BGTK%5D%20) and post link here?

**Comment by clopez** [comment]
Can you try this?

1. Close all nuvolaplayer3 instances.
2. Open a terminal and type:
```
    export WEBKIT_DISABLE_COMPOSITING_MODE=1
    nuvolaplayer3
```
Questions:

1. Is the issue still reproducible after exporting that environment variable?
2. Can you post the output of the following command from a terminal?:
```
    # install package mesa-utils
    glxinfo
```

**Comment by CrazyLemon** [comment]
@clopez thanks for taking a look at this
Answers:
1. It is not. No freeze at all.
2. Sure -> https://paste.ubuntu.com/23047340/
3. The only suspicious thing is maybe this line in dmesg

> [ 1278.071809] traps: WebKitWebProces[4953] general protection ip:7fbb09361414 sp:7ffffc412b10 error:0 in libjavascriptcoregtk-4.0.so.18.3.10[7fbb08b99000+d46000]
> [ 1667.724908] traps: WebKitWebProces[8470] general protection ip:7f3eb2d44414 sp:7ffc354f97b0 error:0 in libjavascriptcoregtk-4.0.so.18.3.10[7f3eb257c000+d46000]

@fenryxo 
I still think there's something wrong with the soundcloud part of nuvolaplayer3.
Lots of criticals and warnings.
Criticals like 

> Worker: [CRITICAL Nuvola] nuvola_js_executor_call_function: assertion 'self != NULL' failed
> Worker: [CRITICAL Nuvola] Application will not function properly.
in thousands of lines.

A few 

> Worker: [CRITICAL  GLib] g_hash_table_insert_internal: assertion 'hash_table != NULL' failed
> Worker: [CRITICAL  GLib] Application will not function properly.

and

> Runner: [WARNING  Nuvola] AppRunnerController.vala:705: Communication failed: Failed to connect to socket '/tmp/libdiorite.nuvolaplayer3.soundcloud.webworker(user)'. Connection refused

When ^ happens the results is usually white screen on the player untill i hit refresh. After refresh i get those criticals. Which keep repeating after X seconds, mouse movements, mouse clicks and so on.

**Comment by clopez** [comment]
I think this may be the WebKitGTK+ bug https://bugs.webkit.org/show_bug.cgi?id=126122

To confirm that, can you install the epiphany browser (GNOME Web), and try loading this test page (warning, it may crash your computer).

https://people.igalia.com/clopez/wkbug/126122/rnyc/

If it crashes or you see a huge amount of memory usage by the WebKitWebProcess process, then its the same issue.

The workaround for now is to disable accelerated/compositing mode by exporting that environment variable.

On the next stable release (2.14) of WebKitGTK+ we are switching to a new threaded compositor that is believed to be free of this bug. The current unstable release (2.13.4) is already using this new threaded compositor by default, just in case you want to test it.

**Comment by CrazyLemon** [comment]
Yup that link definitely crashed my pc.. and this time no ssh!

My suggestion is that you edit your comment and delete that link so people won't go crashing their computers :) 
Do you guys provide a PPA for your releases so i can test it?

**Comment by clopez** [comment]
We don't provide PPAs, sorry. Only source tarballs.

That said, probably the best workaround now for you is to wrap nuvolaplayer3 with a shell script that does an export WEBKIT_DISABLE_COMPOSITING_MODE=1 and then calls the real nuvolaplayer3.

Another idea, is that maybe nuvolaplayer3 itself can do this via a setenv() syscall when the webkitgtk version is < 2.13.4.  Check https://webkitgtk.org/reference/webkit2gtk/stable/webkit2gtk-4.0-WebKitVersion.html

**Comment by fenryxo** [comment]
**A workaround for this issue has landed n the [master](https://github.com/tiliado/nuvolaplayer) & [3.0.x](https://github.com/tiliado/nuvolaplayer/tree/3.0.x) git branches. Thanks for your bug report.**

At first, this fix will be available with the next rebuilt of packages of the [Nuvola Player 3.1 Rolling Releases](https://tiliado.github.io/nuvolaplayer/documentation/3.1/install.html). Later, the fix will be also included in the next bug-fix release of the [Nuvola Player 3.0 Stable Releases](https://tiliado.github.io/nuvolaplayer/documentation/3.0/install.html). To get informed when Nuvola Player 3.0.4 Stable Release is released, follow Nuvola Player on [Facebook](https://www.facebook.com/nuvolaplayer), [Google+](https://plus.google.com/110794636546911932554) or [Twitter](https://twitter.com/NuvolaPlayer) or subscribe to the Nuvola Player Newsletter: [weekly](http://eepurl.com/bLbm5H) or [monthly](http://eepurl.com/bLbtM1).

---

Jiri Janousek, Nuvola Player developer • _My work on Nuvola Player and user support are funded by Nuvola Patrons. ❤ Make a donation to keep the project going. Thank you._ https://tiliado.eu/nuvolaplayer/funding/

**Comment by fenryxo** [comment]
**Update:** This issue has been fixed in the [Nuvola Player 3.0.4 Stable Releases](https://tiliado.github.io/nuvolaplayer/documentation/3.0/install.html) (September 17th, 2016).
